python/pyiceberg/manifest.py

```diff
@@ -76,137 +66,283 @@ def __repr__(self) -> str:
+
+    def fetch_manifest_entry(self, io: FileIO, format_version: int) -> List[ManifestEntry]:
         file = io.new_input(self.manifest_path)
-        return list(read_manifest_entry(file))
-
-
-def read_manifest_entry(input_file: InputFile) -> Iterator[ManifestEntry]:
+        return read_manifest_entry(file, format_version)
+
+
+class ManifestFileV1(ManifestFile):
+    def __setitem__(self, pos: int, value: Any) -> None:
+        if pos == 0:
+            self.manifest_path = value
+        elif pos == 1:
+            self.manifest_length = value
+        elif pos == 2:
+            self.partition_spec_id = value
+        elif pos == 3:
+            self.added_snapshot_id = value
+        elif pos == 4:
+            self.added_files_count = value
+        elif pos == 5:
+            self.existing_files_count = value
+        elif pos == 6:
+            self.deleted_files_count = value
+        elif pos == 7:
+            self.partitions = [
+                element if isinstance(element, PartitionFieldSummary) else PartitionFieldSummary(*element) for element in value
+            ]
+        elif pos == 8:
+            self.added_rows_count = value
+        elif pos == 9:
+            self.existing_rows_count = value
+        elif pos == 10:
+            self.deleted_rows_count = value
+        # ignore the object, it must be from a newer version of the format
+
+
+class ManifestFileV2(ManifestFile):
+    def __setitem__(self, pos: int, value: Any) -> None:
+        if pos == 0:
+            self.manifest_path = value
+        elif pos == 1:
+            self.manifest_length = value
+        elif pos == 2:
+            self.partition_spec_id = value
+        elif pos == 3:
+            self.content = ManifestContent(value)
+        elif pos == 4:
+            self.sequence_number = value
+        elif pos == 5:
+            self.min_sequence_number = value
+        elif pos == 6:
+            self.added_snapshot_id = value
+        elif pos == 7:
+            self.added_files_count = value
+        elif pos == 8:
+            self.existing_files_count = value
+        elif pos == 9:
+            self.deleted_files_count = value
+        elif pos == 10:
+            self.added_rows_count = value
+        elif pos == 11:
+            self.existing_rows_count = value
+        elif pos == 12:
+            self.deleted_rows_count = value
+        elif pos == 13:
+            self.partitions = value
+        elif pos == 14:
+            self.key_metadata = value
+        # ignore the object, it must be from a newer version of the format
+
+
+def read_manifest_entry(input_file: InputFile, format_version: int) -> List[ManifestEntry]:
     with AvroFile(input_file) as reader:
-        schema = reader.schema
-        for record in reader:
-            dict_repr = _convert_pos_to_dict(schema, record)
-            yield ManifestEntry(**dict_repr)
+        return [ManifestEntry.from_record(record, format_version) for record in reader]
```

Comment:
Rather than passing the manifest version down everywhere, I think this should project based on the file schema and a requested schema. All metadata files are maintained so you can always read based on the latest schema (just like any table data file). Plus we want to be able to handle projection.